index.js

```diff
@@ -95,28 +103,34 @@ function flushSeries(collector, seriesName, points) {
     var batch = points.splice(0, spliceIndex);
     var opt = { precision: collector._time_precision };
 
+    collector._flushesInFlight++;
     collector._client.writePoints(seriesName, batch, opt, function(err) {
+        collector._flushesInFlight--;
         if (err) {
             // TODO if error put points back to send again?
-            return self.emit('error', err);
+            collector.emit('error', err);
+            notifyIfFlushed(collector, callback, err);
+            return;
         }
 
         // there are more points to flush out
         if (points.length > 0) {
             flushSeries(collector, seriesName, points);
         }
+        notifyIfFlushed(collector, callback);
     });
 }
 
-Collector.prototype.flush = function() {
+Collector.prototype.flush = function(callback) {
     var self = this;
 
     Object.keys(self._series).forEach(function(key) {
         var series = self._series[key];
         delete self._series[key];
 
-        flushSeries(self, key, series);
+        flushSeries(self, key, series, callback);
```

Comment:
Nit: How do you feel about making these functions be on the prototype? Passing `self` as the first argument is a very C-like pattern that we probably don't need :)